Add support for Offline HIBP password checker

c4rlo opened the pull request:
The new `keepassxc-cli `~`hibp`~`analyze` subcommand can check all passwords in the given database against a given list of SHA-1 password hashes (when `--hibp` is given). Such lists are available from the "Have I Been Pwned" (HIBP) project at https://haveibeenpwned.com/Passwords.

Example usage:

```
$ curl -Lo hashes.7z https://downloads.pwnedpasswords.com/passwords/pwned-passwords-sha1-ordered-by-hash-v4.7z
$ keepassxc-cli analyze --hibp <(7z x -so hashes.7z) passwords.kdbx
Insert password to unlock passwords.kdbx:
Password for entry 'foo' has been leaked 123 times!
Password for entry 'mygroup > bar' has been leaked 789 times!
Finished.
```

[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )

[Have I Been Pwned](https://haveibeenpwned.com/) is a widely used collection of leaked credentials. It is used by several other password managers.

There are several other variations of the functionality implemented in this PR that may be worth considering:

- Ability to access the functionality implemented in this PR from the GUI.
- Ability to access this functionality without having to download the entire password hash database (9.8 GiB compressed currently) first. This is possible using the [HIBP pwned passwords API](https://haveibeenpwned.com/API/v2#PwnedPasswords). To use it, one must send the first 20 bits of the hash up to the server, getting back a list of all matching hashes (generally a few hundred). This is a compromise that leaks a small-ish amount of information about the password to the HIBP server, while keeping data transfer volumes low.
- Ability to check all new or changed passwords against the HIBP API.

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )

The core logic (implemented in `core/HibpOffline.cpp`) has unit tests in `tests/TestHibp.cpp`.

What's still mostly missing is tests for the CLI-specific part.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**

c4rlo commented on src/cli/Command.cpp:
```diff
@@ -75,6 +76,7 @@ void populateCommands()
         commands.insert(QString("estimate"), new Estimate());
         commands.insert(QString("extract"), new Extract());
         commands.insert(QString("generate"), new Generate());
+        commands.insert(QString("hibp"), new Hibp());
```
I'm open to suggestions for better names.

c4rlo commented on src/core/Tools.cpp:
```diff
@@ -116,9 +150,24 @@ namespace Tools
         return formatsStringList.join(" ");
     }
 
+    bool hexDecode(uchar* dest, size_t destSize, const char* src, size_t srcSize)
```
I'm aware of [`QByteArray::fromHex()`](https://doc.qt.io/qt-5/qbytearray.html#fromHex). However, that necessarily results in dynamic memory allocation, which is not ideal for my use case, where I need to do this ~500 million times in a loop.

phoerious replied:
`std::size_t`

c4rlo commented on src/crypto/CryptoHash.h:
```diff
@@ -27,6 +27,7 @@ class CryptoHash
 public:
     enum Algorithm
     {
+        Sha1,
```
This change assumes nothing depends on the order or numeric values of the enum elements. It does not look like anything does, but I wanted to highlight this here anyway just in case.

c4rlo commented on tests/TestCli.cpp:
```diff
@@ -659,6 +661,13 @@ void TestCli::testGenerate()
+{
+    Hibp hibpCmd;
+    QVERIFY(!hibpCmd.name.isEmpty());
+    QVERIFY(hibpCmd.getDescriptionLine().contains(hibpCmd.name));
```
TODO: implement some actual tests here.

louib commented:
@c4rlo thanks for the contribution.

For the record, I opened [a PR](https://github.com/keepassxreboot/keepassxc/pull/1520) for that a few months ago, which I decided to not push through completion. Review comments from that PR might still be relevant (in particular the presence of the SHA1 algo in the standard crypto class).

Your PR would also address https://github.com/keepassxreboot/keepassxc/issues/1083.

One of the popular features currently requested is an integrated password analyzer, that would display information about duplicate passwords, expired passwords, passwords with low entropy, etc. Considering this, we could rename the command so that it allows us to add other metrics / stats. Something like `keepassxc-cli analizer --hibp` ?? Here's the [password analyzer issue](https://github.com/keepassxreboot/keepassxc/issues/551).

c4rlo commented:
Thanks for the context, @louib -- I should have guessed this came up before :)

I'm happy to move the SHA-1 impl out of `CryptoHash` if that will help get this accepted.

As for the name of the subcommand, I'd be happy to go with `analyze --hibp`, especially if we think there will be further analysis modes added in future.

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+
+    struct Sha1Hash
+    {
+        uchar m_data[20];
```
Please do not use char arrays. Use QByteArrays. I don't see why this struct is needed at all. A simple QByteArray should be enough. You are even constructing this object from a QByteArray.

c4rlo replied:
As mentioned [in my comment above](#discussion_r257463912), I did it this way for performance reasons, as I wanted to avoid dynamic memory allocations.

However, I haven't actually benchmarked that so far. I'll do that next, then we can decide based on the results.

phoerious replied:
You do not need to reallocate anything if you reuse the same QByteArray.

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+    uint qHash(const Sha1Hash& b)
+    {
+        uint result;
+        std::memcpy(&result, &b.m_data[0], sizeof(result));
```
All this is not needed with a QByteArray.

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+        return result;
+    }
+
+    bool parseHibpLine(QFile& file, Sha1Hash& sha1, size_t& numPwns, bool& atEnd, QString& error)
```
`error` should be `QString*`.

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+    bool parseHibpLine(QFile& file, Sha1Hash& sha1, size_t& numPwns, bool& atEnd, QString& error)
+    {
+        auto failedTo = [&](const QString& action) {
+            error = QObject::tr("failed to %1").arg(action);
```
Start error messages with capital letters. It's also not worth adding this lambda function just for prepending two words.

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+        };
+
+        auto failedToIo = [&](const QString& action) {
+            error = QObject::tr("failed to %1: %2").arg(action).arg(file.errorString());
```
See above.

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+
+        atEnd = false;
+
+        char hexSha1[sizeof(Sha1Hash::m_data) * 2];
```
Use QByteArray and its size() method (which is O(1) instead of O(n) like sizeof is).

c4rlo replied:
`sizeof` is not O(n). It gets evaluated at compile time.

phoerious replied:
Only true for non-dynamic arrays, which at the time of evaluation you cannot guarantee, since the declaration may change (however unlikely that may be). Generally, do not use plain C arrays.

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+        char hexSha1[sizeof(Sha1Hash::m_data) * 2];
+        qint64 rc = file.read(&hexSha1[0], sizeof(hexSha1));
+        if (rc == -1) {
+            return failedToIo(QObject::tr("read SHA-1"));
```
Make error message a bit more helpful and write it here in full instead of using an extra lambda function (same for all following messages).

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+        return true;
+    }
+
+    QString pwnageReport(const Entry& entry, size_t numPwns)
```
`std::size_t`

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+    QString pwnageReport(const Entry& entry, size_t numPwns)
+    {
+        QStringList ancestors;
+        for (auto g = entry.group(); g && g != g->database()->rootGroup(); g = g->parentGroup()) {
```
A while loop might be easier to understand here. You should also assemble the path string directly. No need to save components to a `QStringList` first and then iterate that again.

c4rlo replied:
The `QStringList` variable is necessary because I need the ancestor path in the reverse order to which I obtain it in this loop.

phoerious replied:
QString has a prepend method. But I guess I'm be fine with the string list approach, too. It may save a few extra copy operations at the cost of more allocations.

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+int Hibp::execute(const QStringList& arguments)
+{
+    TextStream outputTextStream(Utils::STDOUT, QIODevice::WriteOnly);
+    TextStream errorTextStream(Utils::STDERR, QIODevice::WriteOnly);
```
Name them `out` and `err`. I actually don't know why the other modules have the long names again. I renamed them a while ago.

phoerious commented on src/cli/Hibp.cpp:
```diff
@@ -0,0 +1,221 @@
+        }
+
+        for (auto it = entriesBySha1.find(sha1); it != entriesBySha1.end() && it.key() == sha1; ++it) {
+            outputTextStream << pwnageReport(*it.value(), numPwns) << endl;
```
Do you really expect to find multiple entries with the same SHA-1 hash?
Also, shouldn't you output only pwned entries? This list could become awfully long.

c4rlo replied:
Whenever two entries have the same password, they'll have the same SHA-1 of course, and I want to output _all_ matching entries here.

And I am only emitting pwned entries -- the `sha1` I'm searching by comes from the HIBP file over whose lines we're iterating over.

phoerious replied:
True, ignore this comment.

phoerious commented on src/core/Tools.cpp:
```diff
@@ -39,6 +40,39 @@
+namespace
+{
+
+    bool hexDecodeNibble(uchar* result, char c)
```
This function should be indicated as inline.

phoerious commented on src/core/Tools.cpp:
```diff
@@ -30,6 +30,7 @@
 #include <QStringList>
 #include <QUuid>
 #include <cctype>
+#include <iostream>
```
Remove `<iostream>` include.

phoerious commented on src/core/Tools.cpp:
```diff
@@ -39,6 +40,39 @@
+            *result = c - 'A' + 10;
+            return true;
+        } else {
+            std::cerr << "failed to hex-decode bad char '" << c << "'" << std::endl;
```
Do not use `std::cerr`. Use `TextStream` if you must or `qWarning()` if it's not something the user should ever see. In fact, an assertion would be appropriate here.

phoerious commented on src/core/Tools.cpp:
```diff
@@ -39,6 +40,39 @@
+            *result = (n1 << 4) | n2;
+            return true;
+        } else {
+            return false;
```
`else` is redundant.

phoerious commented on src/core/Tools.cpp:
```diff
@@ -116,9 +150,24 @@ namespace Tools
+            return false;
+        }
+
+        for (size_t i = 0; i < srcSize; i += 2, ++dest) {
```
`std::size_t`

phoerious commented on src/crypto/Crypto.cpp:
```diff
@@ -126,6 +126,19 @@ void Crypto::raiseError(const QString& str)
+    QByteArray sha1Test =
+        CryptoHash::hash("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq", CryptoHash::Sha1);
+
+    if (sha1Test != QByteArray::fromHex("84983E441C3BD26EBAAE4AA1F95129E5E54670F1")) {
```
Is this a reference hash from the specification or this or is it just a hash you generated with another tool? You should use and reference the former if possible.

droidmonkey commented:
Recommend not searching the Recycle Bin for passwords

c4rlo commented:
I believe I've addressed most comments now; see individual commits. Still left to do:

- Look into removing `class Sha1Hash` with its `uchar m_data[20]` and just using `QByteArray` instead. I might run some benchmarks first.
- Improve CLI unit tests (in addition to the lower-level unit tests we now have for the underlying logic).

c4rlo commented:
Completed above items. This PR is ready for another round of review.

Btw TeamCity seems to have some issues...

droidmonkey commented on src/core/HibpOffline.cpp:
```diff
@@ -0,0 +1,115 @@
+
+    ParseResult parseHibpLine(QIODevice& input, QByteArray& sha1, std::size_t& numPwns)
+    {
+        char hexSha1[SHA1_BYTES * 2];
```
Do not use raw char unless ABSOLUTELY necessary. Refactor this PR to use QChar, QString, and/or QByteArray in all places.

droidmonkey replied:
After reading your above responses... optimization is not an excuse to use fixed array sizes. This operation only runs when the user requests it and does not require micro-optimizations to save an extra 50 milliseconds.

c4rlo replied:
I looked into using the [`QIODevice::read()`](https://doc.qt.io/qt-5/qiodevice.html#read-1) overload that returns a `QByteArray`. However, as its docs say:

> This function has no way of reporting errors

Therefore, I can no longer distinguish between EOF and I/O error, which is a problem.

If we want to just avoid `char` at all costs, I could of course do:

```c++
QStringArray hexSha1(SHA1_BYTES * 2, '\0');
const qint64 rc = input.read(hexSha1.data(), hexSha1.size());
```

I fail to see how that is any better, since it's just as vulnerable to a buffer overrun if I accidentally pass in the wrong size as the second parameter of the `input.read()` call (which I assume is your concern). But I can do it if required -- let me know.

droidmonkey replied:
It is fine, after chatting with the other team members we can proceed with it as-is.

droidmonkey commented:
I completely missed that this requires to download the entire 9.8 GB password hash database. I find this to be completely untenable for the average user. I would much rather implement the online version instead of this.

c4rlo commented:
I definitely think that supporting an online check would be very nice as well -- certainly more convenient in most cases. But the offline check has a few advantages also:

* Does not require Internet access, so works for people running keepassxc on an airgapped computer.
* Does not leak any information about the passwords. I don't really want to get into a discussion of whether leaking a few bits of a SHA-1 to HIBP is a problem or not, but some people may certainly feel uncomfortable with it.

As regards having to download a 9.8 GB file, I don't see that as a big problem as it's not something that would have to be done often.

droidmonkey commented:
Ok, I'll do a deep dive on this pr once I'm satisfied with 2.4.1

t4777sd commented:
For me personally I would like to have the option to download the data and run it offline.

droidmonkey commented:
I made the necessary updates to merge this PR. Works great, but the HIBP file is HUGE (24 GB uncompressed) and takes about 5 minutes to crunch through on my laptop.
